dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchServiceCollectionExtensions.cs

```diff
@@ -9,260 +9,327 @@
+    [RequiresUnreferencedCode(DynamicCodeMessage)]
+    [RequiresDynamicCode(UnreferencedCodeMessage)]
+    public static IServiceCollection AddAzureAISearchVectorStore(
+        this IServiceCollection services,
+        Func<IServiceProvider, SearchIndexClient>? clientProvider = default,
+        Func<IServiceProvider, AzureAISearchVectorStoreOptions>? optionsProvider = default,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
+        => AddKeyedAzureAISearchVectorStore(services, serviceKey: null, clientProvider, optionsProvider, lifetime);
+
     /// <summary>
-    /// Register an Azure AI Search <see cref="VectorStore"/> with the specified service ID and where <see cref="SearchIndexClient"/> is retrieved from the dependency injection container.
+    /// Registers a keyed <see cref="AzureAISearchVectorStore"/> as <see cref="VectorStore"/>
+    /// with <see cref="SearchIndexClient"/> returned by <paramref name="clientProvider"/>
+    /// or retrieved from the dependency injection container if <paramref name="clientProvider"/> was not provided.
     /// </summary>
-    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="VectorStore"/> on.</param>
-    /// <param name="options">Optional options to further configure the <see cref="VectorStore"/>.</param>
-    /// <param name="serviceId">An optional service id to use as the service key.</param>
-    /// <returns>The service collection.</returns>
-    [RequiresUnreferencedCode("The Azure AI Search provider is currently incompatible with trimming.")]
-    [RequiresDynamicCode("The Azure AI Search provider is currently incompatible with NativeAOT.")]
-    public static IServiceCollection AddAzureAISearchVectorStore(this IServiceCollection services, AzureAISearchVectorStoreOptions? options = default, string? serviceId = default)
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="AzureAISearchVectorStore"/> on.</param>
+    /// <param name="serviceKey">The key with which to associate the vector store.</param>
+    /// <param name="clientProvider">The <see cref="SearchIndexClient"/> provider.</param>
+    /// <param name="optionsProvider">Options provider to further configure the <see cref="AzureAISearchVectorStore"/>.</param>
+    /// <param name="lifetime">The service lifetime for the store. Defaults to <see cref="ServiceLifetime.Singleton"/>.</param>
+    /// <returns>Service collection.</returns>
+    [RequiresUnreferencedCode(DynamicCodeMessage)]
+    [RequiresDynamicCode(UnreferencedCodeMessage)]
+    public static IServiceCollection AddKeyedAzureAISearchVectorStore(
+        this IServiceCollection services,
+        object? serviceKey,
+        Func<IServiceProvider, SearchIndexClient>? clientProvider = default,
+        Func<IServiceProvider, AzureAISearchVectorStoreOptions>? optionsProvider = default,
```

Comment:
this is the most flexible overload: the user can read anything form the service provider to create `SearchIndexClient` and/or `AzureAISearchVectorStoreOptions` or do nothing and we are going to try to resolve it from the DI